### PR TITLE
add CI through GitHub action

### PR DESCRIPTION
**Pull request by junghans**
@dalg24: tested before here: https://github.com/junghans/kokkos/runs/1408542444

There are some segfault using clang on openSUSE, but unrelated to this PR.

### CODE REVIEW

**Comment by masterleinad** [comment]
We should probably figure out why there are segfaults for the failing builds before merging but otherwise, I like the direction of this pull request (in particular testing with ICC).

**Comment by junghans** [comment]
openSUSE is marked as allow to fail for now.

**Comment by junghans** [comment]
@masterleinad @dalg24 here is the backtrace from one of the tests:
```
Reading symbols from /home/junghans/kokkos/build/core/unit_test/KokkosCore_UnitTest_OpenMP...
(gdb) run
Starting program: /home/junghans/kokkos/build/core/unit_test/KokkosCore_UnitTest_OpenMP
Missing separate debuginfos, use: zypper install glibc-debuginfo-2.32-3.1.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7b71e4c in __kmp_fork_call () from /usr/lib64/libomp.so
Missing separate debuginfos, use: zypper install libgcc_s1-debuginfo-10.2.1+git872-1.1.x86_64 libhwloc15-debuginfo-2.2.0-1.2.x86_64 liblzma5-debuginfo-5.2.5-1.16.x86_64 libomp11-devel-debuginfo-11.0.0-1.2.x86_64 libpciaccess0-debuginfo-0.16-1.5.x86_64 libstdc++6-debuginfo-10.2.1+git872-1.1.x86_64 libz1-debuginfo-1.2.11-16.1.x86_64
(gdb) bt
#0  0x00007ffff7b71e4c in __kmp_fork_call () from /usr/lib64/libomp.so
#1  0x00007ffff7b5fcc2 in __kmpc_fork_call () from /usr/lib64/libomp.so
#2  0x00000000015be419 in Kokkos::OpenMP::impl_initialize(int) ()
#3  0x00000000015beab1 in Kokkos::Impl::OpenMPSpaceInitializer::initialize(Kokkos::InitArguments const&) ()
#4  0x00000000015abe37 in Kokkos::Impl::ExecSpaceManager::initialize_spaces(Kokkos::InitArguments const&) ()
#5  0x00000000015afedd in Kokkos::Impl::(anonymous namespace)::initialize_backends(Kokkos::InitArguments const&) ()
#6  0x00000000015af4de in Kokkos::Impl::(anonymous namespace)::initialize_internal(Kokkos::InitArguments const&) ()
#7  0x00000000015accbd in Kokkos::initialize(int&, char**) ()
#8  0x0000000000415993 in main ()
```
this is with clang-11.

Conveniently you build kokkos inside the `ghcr.io/kokkos/ci-containers/opensuse:latest` container to reproduce this.

**Comment by junghans** [comment]
I updated the backtrace above with all symbols.

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:12]:
```suggestion
  schedule:
    - cron:  '0 5 * * FRI'
```
Any reason for doing a scheduled run?

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:19]: This looks a bit excessive for CI. If at all, I would test multiple distros only on merge commits (and not on every commit in a pull request) as a regression test.

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:16]:
```suggestion
# FIXME
    continue-on-error: ${{ matrix.distro == 'opensuse:latest' }}
```
All other distros work, don't they?

**Review by masterleinad — COMMENTED** [review]
I would be fine with merging this after addressing my comments. In the end, we can only judge properly the usefulness when seeing it in action a little bit.
In my opinion, having a build with `icpc` is the most compelling reason for moving forward here.

File: .github/workflows/continuous-integration-workflow.yml
Comment (junghans) [reply .github/workflows/continuous-integration-workflow.yml:12]: Nothing in particular, just so get a test run at least once a week in case the containers changed.

File: .github/workflows/continuous-integration-workflow.yml
Comment (junghans) [reply .github/workflows/continuous-integration-workflow.yml:16]: `fedora:rawhide` and `ubuntu:devel` are a bit flacky at times hence I added them here.

File: .github/workflows/continuous-integration-workflow.yml
Comment (junghans) [reply .github/workflows/continuous-integration-workflow.yml:19]: The builds are relatively quick, so I didn't do a pre-selection, but we can do whatever. To build on different distros for PRs vs. pushes on master, we would need to split this into 2 files. Fine with me, just saying.

File: .github/workflows/continuous-integration-workflow.yml
Comment (junghans) [reply .github/workflows/continuous-integration-workflow.yml]: Removed.

**Comment by junghans** [comment]
Should I remove `.travis.yml` in the PR as well?

BTW, the containers used are generated here: https://github.com/kokkos/ci-containers

File: .github/workflows/continuous-integration-workflow.yml
Comment (junghans) [reply .github/workflows/continuous-integration-workflow.yml:17]: 56 builds seem not overburdensome, but it is your call.

**Comment by masterleinad** [comment]
Hmmm... apparently the failing builds are still shown as breaking. Does `continue-on-error` not work as expected? 
The selection of OS looks unusual but I don't feel really strongly about it.

**Comment by junghans** [comment]
> Hmmm... apparently the failing builds are still shown as breaking. Does `continue-on-error` not work as expected?
> The selection of OS looks unusual but I don't feel really strongly about it.

`continue-on-error` property just means don't cancel the other builds when this build fails. It won't overwrite the exit status though.

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:14]:
```suggestion
    continue-on-error: true
```
I think it makes sense to always run all the tests.

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:17]:
```suggestion
        distro: ['fedora:latest', 'fedora:rawhide', 'ubuntu:latest']
```

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [inline .github/workflows/continuous-integration-workflow.yml:10]:
```suggestion
on: [push, pull_request]
```
Let's be consistent with the OSX check here.

File: .github/workflows/continuous-integration-workflow.yml
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:17]: This should cover current CI builds.

**Review by masterleinad — CHANGES_REQUESTED** [review]
I briefly tried to reproduce and debug the failing `opensuse` builds but they were passing for me building the container locally.
This is certainly an issue to look into but I would prefer to have something reasonable to move forward so I would replace it with `ubuntu` for now. We can always refine later.

**Comment by masterleinad** [comment]
This is passing now.

File: .github/workflows/continuous-integration-workflow.yml
Comment (dalg24) [inline .github/workflows/continuous-integration-workflow.yml:1]:
```suggestion
name: github-Linux
```
for consistency with "github-OSX" or similar because the checks currently read "CI / CI"

File: .github/workflows/continuous-integration-workflow.yml
Comment (DavidPoliakoff) [inline .github/workflows/continuous-integration-workflow.yml:32]: "trial" I think (excuse the nit)

**Comment by DavidPoliakoff** [comment]
I'm not well versed in CI, so these comments aren't "I think something is wrong," they're "help me understand." Why do I care about running on Fedora rawhide and the other Fedora?

File: .github/workflows/continuous-integration-workflow.yml
Comment (dalg24) [reply .github/workflows/continuous-integration-workflow.yml:32]:
```suggestion
      - name: Get trial license
```

**Comment by DavidPoliakoff** [comment]
We're discussing on Slack. Nobody has a _problem_ with doing a little extra, so we can discuss my last comment if we ever hit the point where we want to cut back on builds

**Review by dalg24 — APPROVED** [review]
We will merge as is and re-evaluate as needed.
Thanks for working on this

**Review by DavidPoliakoff — APPROVED** [review]

**Pull request merged**
